### PR TITLE
Histogram

### PR DESCRIPTION
**Pull request by shkr**
I have implemented a Histogram chart using design from http://bl.ocks.org/mbostock/1933560. 
It extends BaseViz. It draws a histogram using d3.layout.histogram method after querying a given numeric column from the db. It changes color for different bar lengths in the histogram. It takes as input the no of bins for the histogram.

histogram
 (Updated Thumbnail!)

### CODE REVIEW

**Comment by shkr** [comment]
In response to Issue #888 posted and commented on earlier

File: caravel/viz.py [hunk 1]
```diff
@@ -7,7 +7,6 @@
 from __future__ import division
 from __future__ import print_function
 from __future__ import unicode_literals
-
```
Comment (xrmx) [inline caravel/viz.py]: unrelated change

File: caravel/viz.py [hunk 2]
```diff
@@ -47,8 +46,7 @@ class BaseViz(object):
         'label': None,
         'fields': (
             'metrics', 'groupby',
-        )
-    },)
+        ) },)
```
Comment (xrmx) [inline caravel/viz.py]: same here

File: caravel/viz.py [hunk 3]
```diff
@@ -1200,6 +1198,87 @@ def get_data(self):
...
+class HistogramViz(BaseViz):
+    """Histogram"""
+
+    viz_type = "histogram"
+    verbose_name = _("Histogram")
+    is_timeseries = False
+    fieldsets = ({
+        'label': None,
+        'fields': (
+            ('all_columns_x'),
```
Comment (xrmx) [inline caravel/viz.py]: Don't you need a comma here to make it a tuple?

File: caravel/assets/visualizations/histogram.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        gEnter.append("g").attr("class", "x axis");
+
+        svg.attr("width", slice.width())
+           .attr("height", slice.height());
+        
+        var g = svg.select("g")
+            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
+        
+        function colors(n) {         
+            var color_list = [
```
Comment (xrmx) [inline caravel/assets/visualizations/histogram.js]: The other modules usually pick a color from px.color.category21 from javascripts/modules/caravel.js.

Comment (shkr) [inline caravel/assets/visualizations/histogram.js]: Thanks. I did not know.  I will make the change.

File: caravel/viz.py (hunk 3 above)
Comment (shkr) [inline caravel/viz.py]: Which tuple is missing a trailing comma ?

Comment (xrmx) [inline caravel/viz.py]: This:
```
('all_columns_x'),
```

**Comment by shkr** [comment]
@xrmx i have fixed the issues you posted

File: caravel/assets/visualizations/histogram.css [hunk 5]
```diff
@@ -0,0 +1,24 @@
+.bars rect {}
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.css]: unused style rule, let's delete this

File: caravel/assets/visualizations/histogram.css [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+.axis path
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.css]: unused style rule

File: caravel/assets/visualizations/histogram.css [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+.axis line {
+  fill: none;
+  stroke: black;
+  shape-rendering: crispEdges;
+}
+
+.axis text {
+  font-family: sans-serif;
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.css]: let's use this for declaring font-family: 
`font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif;`

File: caravel/assets/visualizations/histogram.css [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  font-size: 11px;
+}
+
+.axis text {
+    font: 10px sans-serif;
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.css]: `font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif;`

File: caravel/assets/visualizations/histogram.css [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  stroke: black;
+  shape-rendering: crispEdges;
+}
+
+.axis text {
+  font-family: sans-serif;
+  font-size: 11px;
+}
+
+.axis text {
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.css]: this is a duplicate rule, let's delete it please.

File: caravel/assets/visualizations/histogram.css [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+.axis line {
+  fill: none;
+  stroke: black;
+  shape-rendering: crispEdges;
+}
+
+.axis text {
+  font-family: sans-serif;
+  font-size: 11px;
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.css]: can we bump this to 12px, plz

File: caravel/assets/visualizations/histogram.js [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+var px = window.px || require('../javascripts/modules/caravel.js')
+
+// CSS
+require('./histogram.css')
+
+function histogram(slice) {
+    
+    var div = d3.select(slice.selector)
+    
+    var _draw = function(data, noOfBins) {
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.js]: could we change this to `numBins` rather than `noOfBins` for clarity?

File: caravel/assets/visualizations/histogram.js [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+              padding = 12.0
+            } else {
+              padding = -8.0
+            }
+            return y(d.y) - padding;
+          })
+          .attr("x", function(d) { return x(d.x) + (x.rangeBand()/2);})
+          .attr("text-anchor", "middle")
+          .text(function(d) { return formatNumber(d.y); })
+          .attr("fill", "black")
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.js]: i think white text within bar is more readable than black.

screenshot 2016-08-08 12 16 52

File: caravel/assets/visualizations/histogram.js [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+    var render = function() {
+    
+        d3.json(slice.jsonEndpoint(), function(error, json) {
+            if(error !== null) {
+                slice.error(error.responseText, error);
+                return '';
+            }
+            
+            var noOfBins = Number(json.form_data.link_length) || 10;
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.js]: `s/noOfBins/numBins`

File: caravel/assets/visualizations/histogram.js [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+// CSS
+require('./histogram.css')
+
+function histogram(slice) {
+    
+    var div = d3.select(slice.selector)
+    
+    var _draw = function(data, noOfBins) {
+        
+        var margin = { top: 50, right: 10, bottom: 20, left: 50 },
```
Comment (ascott) [inline caravel/assets/visualizations/histogram.js]: we are using the airbnb open source javascript style guide in this code base now, could you update to follow these guidelines? 

https://github.com/airbnb/javascript

a few that jump out: 
- single variable per line 
- use const or let
- use single quotes in js
- use trailing commas

we have an automatic linter happening on all js EXCEPT visualizations, but it would be great for all new visualizations to use this code style. thx!

Comment (shkr) [inline caravel/assets/visualizations/histogram.js]: Got it. I will make the change.

**Comment by shkr** [comment]
@ascott made the changes you recommended

**Comment by mistercrunch** [comment]
Neat. The thumbnail image should be square and 1024*1024 I believe.

**Comment by shkr** [comment]
histogram

1024x1024 visualization picture updated

**Comment by shkr** [comment]
@mistercrunch I checked out the details for the code-quality all the changes recommended except 1 PEP302 violation, are out of scope for this PR

File: caravel/viz.py [hunk 15]
```diff
@@ -1200,6 +1200,78 @@ def get_data(self):
         df.columns = ['x', 'y']
         return df.to_dict(orient="records")
```
Comment (xrmx) [inline caravel/viz.py]: one more empty line here

File: caravel/viz.py [hunk 16]
```diff
@@ -1200,6 +1200,78 @@ def get_data(self):
...
+        chart_data = []
+        for value in df[df.columns[0]].values:
+            chart_data.append(value)
+        return chart_data
+
+    def get_data(self):
+        """Returns the chart data"""
+        df = self.get_df()
+        chart_data = self.to_series(df)
+        return chart_data
```
Comment (xrmx) [inline caravel/viz.py]: one more empty line here too

File: caravel/viz.py [hunk 17]
```diff
@@ -1200,6 +1200,78 @@ def get_data(self):
...
+        },
+        'link_length': {
+            'label': _("No of Bins"),
+            'description': _("Select number of bins for the histogram"),
+            'default': 5
+        }
+    }
+
+    def query_obj(self):
+
```
Comment (xrmx) [inline caravel/viz.py]: docstring or just remove the line :)

File: caravel/viz.py [hunk 18]
```diff
@@ -1200,6 +1200,78 @@ def get_data(self):
...
+        d = super(HistogramViz, self).query_obj()
+        d['row_limit'] = self.form_data.get('row_limit') or 100000
```
Comment (xrmx) [inline caravel/viz.py]: self.form_data.get('row_limit', int(config.get("ROW_LIMIT")))

File: caravel/viz.py [hunk 19]
```diff
@@ -1200,6 +1200,78 @@ def get_data(self):
...
+        numeric_column = self.form_data.get('all_columns_x') or None
```
Comment (xrmx) [inline caravel/viz.py]: no need for "or None" part

File: caravel/viz.py [hunk 20]
```diff
@@ -1200,6 +1200,78 @@ def get_data(self):
...
+
+        df.replace([np.inf, -np.inf], np.nan)
+        df = df.fillna(0)
+
+        return df
+
+    def to_series(self, df):
+        """Returns a list of numeric values from a dataframes's first column"""
+        chart_data = []
+        for value in df[df.columns[0]].values:
```
Comment (xrmx) [inline caravel/viz.py]: does *return df[df.columns[0]].value.to_list()* returns the same?

Comment (shkr) [inline caravel/viz.py]: Alright. I will make the changes

**Comment by shkr** [comment]
@xrmx Changes made as recommended

**Comment by xrmx** [comment]
@shkr thanks, python side looks good to me!

**Comment by mistercrunch** [comment]
LGTM mergin'

**Comment by ascott** [comment]
nice work, thx @shkr!
